From bca6e054261b4dba0889117e9520d478988bba17 Mon Sep 17 00:00:00 2001
From: Brian Ryner <bryner@google.com>
Date: Thu, 27 Aug 2020 11:42:01 +1000
Subject: [PATCH 3/3] For MacOS, use the code path that counts file descriptors
 under /dev/fd rather than trying to dup() all possible FDs. The latter method
 does not work reliably under MacOS 11.

---
 Lib/test/support/__init__.py | 9 +++++----
 1 file changed, 5 insertions(+), 4 deletions(-)

diff --git a/Lib/test/support/__init__.py b/Lib/test/support/__init__.py
index ccc11c1b4b..ffc252a6f6 100644
--- a/Lib/test/support/__init__.py
+++ b/Lib/test/support/__init__.py
@@ -2081,11 +2081,12 @@ def _crash_python():
 def fd_count():
     """Count the number of open file descriptors.
     """
-    if sys.platform.startswith(('linux', 'freebsd')):
+    if sys.platform.startswith(('linux', 'freebsd', 'darwin')):
         try:
-            names = os.listdir("/proc/self/fd")
-            # Substract one because listdir() opens internally a file
-            # descriptor to list the content of the /proc/self/fd/ directory.
+            fd_dir = '/dev/fd' if sys.platform == 'darwin' else '/proc/self/fd'
+            names = os.listdir(fd_dir)
+            # Subtract one because listdir() opens internally a file descriptor
+            # to list the contents of fd_dir.
             return len(names) - 1
         except OSError as exc:
             if exc.errno != errno.ENOENT:
-- 
2.28.0.297.g1956fa8f8d-goog

